README.md
# Numbers to words

This package allows you to easily convert numbers to words.

## Usage

[Example](https://github.com/champ96k/int_to_words/tree/master/example)

To use this package :

- add the dependency to your pubspec.yaml file.

```yaml
dependencies:
  flutter:
    sdk: flutter
  int_to_words: ^0.0.4
```

### How to use

```dart
  final NumberToWords _number = NumberToWords();

  final _num = _number.convert(022); -> Twenty Two

  final _num1 = _number.convert(47474747); -> Forty Seven Million Four Hundred and Seventy Four Thousand Seven Hundred and Forty Seven

 
```

### Created & Maintained By

[Tushar Nikam](https://champ96k.github.io)

<a href="https://www.twitter.com/champ_96k"><img src="https://img.shields.io/badge/twitter-%231DA1F2.svg?&style=for-the-badge&logo=twitter&logoColor=white" height=25> </a>
<br>
<a href="https://www.linkedin.com/in/tushar-nikam-a29a97131/"><img src="https://img.shields.io/badge/linkedin-%230077B5.svg?&style=for-the-badge&logo=linkedin&logoColor=white" height=25></a>


CHANGELOG.md
## 0.0.1

* A Flutter package to easily convert numbers to words.


## 0.0.2

* code refactor + documentation added


## 0.0.3

* code refactor + documentation added

## 0.0.4

* code refactor + documentation added

test/int_to_words_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:int_to_words/int_to_words.dart';

void main() {
  test('Converts number to words', () {
    final _numberToWords = IntToWords();
    expect(_numberToWords.convert(10), "Ten");
    expect(_numberToWords.convert(9999999999),
        "Nine Thousand  Nine Hundred and Ninety Nine Million  Nine Hundred and Ninety Nine Thousand Nine Hundred and Ninety Nine");
  });
}


test/number_to_words_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:numbers_to_word/numbers_to_word.dart';

void main() {
  test('Converts number to words', () {
    final _numberToWords = NumbersToWord();
    expect(_numberToWords.convert(10), "Ten");
    expect(_numberToWords.convert(9999999999),
        "Nine Thousand  Nine Hundred and Ninety Nine Million  Nine Hundred and Ninety Nine Thousand Nine Hundred and Ninety Nine");
  });
}


lib/number_to_words.dart
class NumberToWords {

 final List<String> _units = [
    "Zero",
    "One",
    "Two",
    "Three",
    "Four",
    "Five",
    "Six",
    "Seven",
    "Eight",
    "Nine",
    "Ten",
    "Eleven",
    "Twelve",
    "Thirteen",
    "Fourteen",
    "Fifteen",
    "Sixteen",
    "Seventeen",
    "Eighteen",
    "Nineteen"
  ];
 final List<String> _tens = [
    "",
    "",
    "Twenty",
    "Thirty",
    "Forty",
    "Fifty",
    "Sixty",
    "Seventy",
    "Eighty",
    "Ninety"
  ];

  String _convert(num i) {
    /// 0 -> 20
    if (i < 20) {
      return _units[i.toInt()];
    }
    /// 20 -> 100
   else if (i < 100){
      return _tens[i ~/ 10] + ((i % 10 > 0) ? " " + _convert(i % 10) : "");
    }

    /// 100 -> 1000
    else if (i < 1000) {
      return _units[i ~/ 100] +
          " Hundred" +
          ((i % 100 > 0) ? " and " + _convert(i % 100) : "");
    }

    /// 1000 -> 1000000
    else if (i < 1000000) {
      return _convert(i / 1000) +
          " Thousand " +
          ((i % 1000 > 0) ? " " + _convert(i % 1000) : "");
    }
    /// 1000000 onwards
    else {
      return _convert(i / 1000000) +
          " Million " +
          ((i % 1000000 > 0) ? " " + _convert(i % 1000000) : "");
    }


  }

   String convert(num i) {
    final _result = _convert(i);
    return _result.replaceAll(RegExp('\\s+'), ' ');
  }
}


lib/numbers_to_word.dart
class NumbersToWord {
  final List<String> _units = [
    "Zero",
    "One",
    "Two",
    "Three",
    "Four",
    "Five",
    "Six",
    "Seven",
    "Eight",
    "Nine",
    "Ten",
    "Eleven",
    "Twelve",
    "Thirteen",
    "Fourteen",
    "Fifteen",
    "Sixteen",
    "Seventeen",
    "Eighteen",
    "Nineteen"
  ];
  final List<String> _tens = [
    "",
    "",
    "Twenty",
    "Thirty",
    "Forty",
    "Fifty",
    "Sixty",
    "Seventy",
    "Eighty",
    "Ninety"
  ];

  String _convert(num i) {
    /// 0 -> 20
    if (i < 20) {
      return _units[i.toInt()];
    }

    /// 20 -> 100
    else if (i < 100) {
      return _tens[i ~/ 10] + ((i % 10 > 0) ? " " + _convert(i % 10) : "");
    }

    /// 100 -> 1000
    else if (i < 1000) {
      return _units[i ~/ 100] +
          " Hundred" +
          ((i % 100 > 0) ? " and " + _convert(i % 100) : "");
    }

    /// 1000 -> 1000000
    else if (i < 1000000) {
      return _convert(i / 1000) +
          " Thousand " +
          ((i % 1000 > 0) ? " " + _convert(i % 1000) : "");
    }

    /// 1000000 onwards
    else {
      return _convert(i / 1000000) +
          " Million " +
          ((i % 1000000 > 0) ? " " + _convert(i % 1000000) : "");
    }
  }

  String convert(num i) {
    final _result = _convert(i);
    return _result.replaceAll(RegExp('\\s+'), ' ');
  }
}


lib/int_to_words.dart
class IntToWords {
  ///
  final List<String> _units = [
    "Zero",
    "One",
    "Two",
    "Three",
    "Four",
    "Five",
    "Six",
    "Seven",
    "Eight",
    "Nine",
    "Ten",
    "Eleven",
    "Twelve",
    "Thirteen",
    "Fourteen",
    "Fifteen",
    "Sixteen",
    "Seventeen",
    "Eighteen",
    "Nineteen"
  ];

  ///
  final List<String> _tens = [
    "",
    "",
    "Twenty",
    "Thirty",
    "Forty",
    "Fifty",
    "Sixty",
    "Seventy",
    "Eighty",
    "Ninety"
  ];

  ///Recursive [_convert] function to convert [int] to numbers [string]
  String _convert(num i) {
    /// 0 -> 20
    if (i < 20) {
      return _units[i.toInt()];
    }

    /// 20 -> 100
    else if (i < 100) {
      return _tens[i ~/ 10] + ((i % 10 > 0) ? " " + _convert(i % 10) : "");
    }

    /// 100 -> 1000
    else if (i < 1000) {
      return _units[i ~/ 100] +
          " Hundred" +
          ((i % 100 > 0) ? " and " + _convert(i % 100) : "");
    }

    /// 1000 -> 1000000
    else if (i < 1000000) {
      return _convert(i / 1000) +
          " Thousand " +
          ((i % 1000 > 0) ? " " + _convert(i % 1000) : "");
    }

    /// 1000000 onwards
    else {
      return _convert(i / 1000000) +
          " Million " +
          ((i % 1000000 > 0) ? " " + _convert(i % 1000000) : "");
    }
  }

  String convert(num i) {
    final _result = _convert(i);

    ///return string result without duplicated spaces
    return _result.replaceAll(RegExp('\\s+'), ' ');
  }
}
